### PR TITLE
Lh cd 287

### PR DESCRIPTION
**Pull request by lily-hemmett-rl-cix**
Add function to get day of week transaction counts

### CODE REVIEW

File: rlutils/features/calc/customer.py [hunk 1]
```diff
@@ -181,3 +182,42 @@ def max_value_tran(df: pd.DataFrame,
...
+    Returns:
+        pd.DataFrame: A customer business unit level dataframe containing the count of unique transactions made on a given day of the week.
+    """
+    if filt_returns: df = filter_returns(df, transaction_type_col)
+    df = filter_by_date(df, start_date, end_date, date_col)
+
+    if not is_datetime(df[date_col]): df[date_col] = pd.to_datetime(df[date_col])
+
+    tran_dates = df.groupby([cust_id_col, bu_col, date_col], as_index=False)[tran_id_col].nunique()
+    tran_dates['dow'] = tran_dates[date_col].dt.day_name().apply(lambda x: x[:3])
```
Comment (andrew-boney-rl-cix) [inline rlutils/features/calc/customer.py:217]: I was going to suggest changing this to
```
tran_dates[date_col].dt.day_name().str[:3]
```
but I tested it locally it turns out your way is significantly faster
image

so lets stick with your way

File: rlutils/features/calc/customer.py [hunk 2]
```diff
@@ -181,3 +182,42 @@ def max_value_tran(df: pd.DataFrame,
...
+    tran_dates_piv = pd.pivot_table(tran_dates, index=[cust_id_col, bu_col], columns='dow', values=tran_id_col, aggfunc='sum', fill_value=0).reset_index()
+    tran_dates_piv.columns = [cust_id_col, bu_col] + [x+"_tran_count" for x in tran_dates_piv.columns if x not in [cust_id_col, bu_col]]
+    missing_cols = [y for y in ["Mon_tran_count", "Tue_tran_count", "Wed_tran_count", "Thu_tran_count", "Fri_tran_count", "Sat_tran_count", "Sun_tran_count"] if y not in [x for x in tran_dates_piv.columns if "tran_count" in x]]
+    for i in missing_cols:
+        tran_dates_piv[i] = 0
```
Comment (andrew-boney-rl-cix) [inline rlutils/features/calc/customer.py:222]: I tried implementing this a bit differently, basically using ``` pd.Categorical ``` which would automatically fill missing values. (which would be faster & more succinct).

However, having difficulties as pandas is stupid and makes it so that if any column in a pivot is categorical, it treats all pivotting columns (i.e. anything in index, values or columns) as categorical which results in all customers having rows for all business units in the input DataFrame.  

I've wasted enough time on this already and am not going anywhere so will merge as is for now. My code is below if I can be bothered to go back and implement later

```
def dow_tran_counts2(df: pd.DataFrame,
                    start_date: str = "1900-01-01",
                    end_date: str = "2100-01-01",
                    cust_id_col: str = "individual_id",
                    bu_col: str = "business_unit_code",
                    tran_id_col: str = "transaction_id",
                    date_col: str = "transaction_datetime",
                    transaction_type_col: str = "transaction_type_code",
                    filt_returns: bool = True
                    ) -> pd.DataFrame:
    """Get transaction counts by day of week

    Args:
        df (pd.DataFrame): transaction dataframe
        start_date (str, optional): (inclusive) Discard transactions before given start date. Defaults to "1900-01-01".
        end_date (str, optional): (inclusive) Discard transactions after given end date. Defaults to "2100-01-01".
        cust_id_col (str, optional): Column with customer ID in. Defaults to "individual_id".
        bu_col (str, optional): Column with business unit ID in. Defaults to "business_unit_code".
        tran_id_col (str, optional): Column with transaction ID in. Defaults to "transaction_id".
        date_col (str, optional): Column with date of transaction in. Defaults to "transaction_datetime".
        transaction_type_col (str, optional): Column identifying sale/returns. Defaults to "transaction_type_code".
        filt_returns (bool, optional): Whether or not to remove returns from the transactions before calculating. Defaults to True.

    Returns:
        pd.DataFrame: A customer business unit level dataframe containing the count of unique transactions made on a given day of the week.
    """
    if filt_returns: df = filter_returns(df, transaction_type_col)
    df = filter_by_date(df, start_date, end_date, date_col)

    if not is_datetime(df[date_col]): df[date_col] = pd.to_datetime(df[date_col])

    tran_dates = df.groupby([cust_id_col, bu_col, date_col], as_index=False)[tran_id_col].nunique()
    tran_dates['dow'] = pd.Categorical(tran_dates[date_col].dt.day_name().apply(lambda x: x[:3]), 
                                        categories = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'])

    tran_dates_wide = (pd.pivot_table(tran_dates, 
                                    index=[cust_id_col, bu_col], columns='dow', values=tran_id_col, 
                                    aggfunc='sum', fill_value=0)
                        .reset_index())

    tran_dates_wide.columns = [cust_id_col, bu_col] + [f"{c}_tran_count" for c in tran_dates["dow"].cat.categories]

    return tran_dates_wide
```

Comment (lily-hemmett-rl-cix) [reply rlutils/features/calc/customer.py:222]: Yeah, I agree this is probably not the best way. I actually didn't even think about it until it failed the unit test! But I guess if you are looking over a very small time set or some other filter it may also arise. Very much an edge case though

Comment (andrew-boney-rl-cix) [reply rlutils/features/calc/customer.py:222]: Yeah, it's definitely needed in one form or another. Would cause some really annoying bugs if not
